Refactor Expr to DXL translator for list partition selector predicates

hsyuan opened the pull request:
Previously when the final plan is translated into DXL, Orca uses range predicates to represent the values of the list partition (range start and end value equal a single list value), which is error-prone and redundant for the query executor.

In this patch, we use the following way to translate the predicates of the list partition selector in DXL (range based partition remain unchanged):
1. PK op Scalar -> Scalar reverse(op) Any(PartListValues)
e.g. 
For partition selector predicate `pk1 < 5`, will be translated to `ArrayComp(5 > Any(PartListValues))`, which means as long as any value of list partition values is less than 5, the partition will be selected.
2. PK is (not) NULL -> PartListNullTest
3. Propagation Expression ->
Const1 = Any(PartListValues) or Const2 = Any(PartListValues) ...

oarap commented:
Is it possible to make the history cleaner and more reviewable? I would initially suggest to put all test file changes into one commit so that the reviewers can focus on the core development.

hsyuan commented:
cleaned commit history as requested.

hsyuan commented:
It is kind of messy PR, I have to admit. IMHO, only 2 files needs your attention: `CTranslatorExprToDXLUtils.cpp` and `CTranslatorExprToDXL.cpp`

d commented on libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -4966,37 +5003,84 @@ CTranslatorExprToDXL::PdxlnScCmpPartKey
 
+	// For varchar(1043) like columns, GPDB will automatically cast to text(25).
+	// So we need special care to get the actual type from the expression.
+	pmdidTypePartKey = CScalar::PopConvert(pexprPartKey->Pop())->PmdidType();
```
No no no!
This is *throwing away* the original type of the partition key, which is not only logically wrong, but also leads to wrong type information in the plan.

Which query motivated this? If we drop this line what problem do you foresee?

hsyuan replied:
as you can see from the comment, any list part table with varchar part key

d replied:
query

d replied:
That's because you missed the cast from the partition key side of the comparison. I added comments inline

hsyuan replied:
correct, that's what I want to avoid.

hsyuan commented on libnaucrates/src/CDXLUtils.cpp:
```diff
@@ -1973,6 +1973,34 @@ CDXLUtils::PstrSerialize
+		}
+	}
+
+	return a_pstr.PtReset();
```
I didn't get it. Can you give an example?

d replied:
Meaning: stop using the smart pointer (`AutoP`)

hsyuan commented on libnaucrates/include/naucrates/md/CMDRelationGPDB.h:
```diff
@@ -131,6 +134,7 @@ namespace gpmd
 				DrgPmdcol *pdrgpmdcol,
 				DrgPul *pdrgpulDistrColumns,
 				DrgPul *pdrgpulPartColumns,
+				DrgPsz *m_pdrgpszPartTypes,
```
nice catch.

hsyuan commented:
The partition type is enum. even though it's specified by char, but it's viewed as int.

d commented on libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -4966,37 +5003,84 @@ CTranslatorExprToDXL::PdxlnScCmpPartKey
-	CTranslatorExprToDXLUtils::ExtractCastMdids(pexprNewPartKey->Pop(), &pmdidTypeCastExpr, &pmdidCastFunc);
+		// If the pexprOther is not comparable with pexprPartKey, but can be casted to pexprPartKey,
+		// and not yet casted, then we add a cast on top of pexprOther.
+		if (!CMDAccessorUtils::FCmpExists(m_pmda, pmdidTypeOther, pmdidTypePartKey, ecmpt))
```
This is wrong for several reasons:

0. It is casting the "other" side instead of the side of the partition key
0. This combined with the check for cast on line 5055, missed the case when the partition key expression is given in the form of a cast, hence will lead to a double cast.

The fix for this is not to introduce a *fake* `pmdidTypePartKey` in line 5008 above, but instead, stay consistent with the range partition case: i.e. we add cast to the partition key side.

hsyuan replied:
Can you give an example that can have double cast? The plan can't be generated, as far as I know.

d replied:
If we remove the fake `pmdidTypePartKey` assignment from line 5008, wouldn't this cast the right hand side (again)?

hsyuan replied:
no, I don't think so.

d commented:
The handling of cast here is wrong. I posted a small repro over at https://github.com/greenplum-db/gpdb/pull/2076#pullrequestreview-29344763 .

In both list and range partitions, there are cases where we will generate a plan that wrongly advertises the value of the partition key (type of `PARTBOUNDEXPR` and element type of `PARTLISTEXCPR`, respectively).

In general, instead of rigging the executor to increasingly relax the expectation of our plan's correctness, we should fix it in ORCA:

hsyuan commented:
Well, if you want to support GPDB 5 only,  I can modify it that way.

Sent from my iPhone

> 在 2017年3月27日，21:34，Jesse Zhang <notifications@github.com> 写道：
> 
> The handling of cast here is wrong. I posted a small repro over at greenplum-db/gpdb#2076 (review) .
> 
> In both list and range partitions, there are cases where we will generate a plan that wrongly advertises the value of the partition key (type of PARTBOUNDEXPR and element type of PARTLISTEXCPR, respectively).
> 
> In general, instead of rigging the executor to increasingly relax the expectation of our plan's correctness, we should fix it in ORCA:
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.
>

d commented:
To recap our discussion offline, array coerce expressions _are portable_ even to GPDB 4. The translator in a database without Array Coerce Expression needs to translate it to a FUNCEXPR with the corresponding `resulttype` and `fnoid`

hsyuan commented:
Updated code to address comments.

bhuvnesh2703 commented on libnaucrates/src/CDXLUtils.cpp:
```diff
@@ -1975,6 +1975,41 @@ CDXLUtils::PstrSerialize
+//	@doc:
+//		Serialize a list of chars into a comma-separated string
+//
+//---------------------------------------------------------------------------
```
could you please remove this type of comment and have 
```
// serialize a list of chars into a comma-separated string
```

hsyuan replied:
you are looking at the old commit

oarap commented on libnaucrates/src/CDXLUtils.cpp:
```diff
@@ -1973,6 +1973,34 @@ CDXLUtils::PstrSerialize
+	CWStringDynamic *pstr = GPOS_NEW(pmp) CWStringDynamic(pmp);
+
+	ULONG ulLength = pdrgsz->UlLength();
+	for (ULONG ul = 0; ul < ulLength; ul++)
```
you can loop until ulLength-1

and after the loop you append it without comma. this would help you avoid if, else

hsyuan replied:
make sense

hsyuan replied:
well, if changing it this way, the code going to be:
```cpp
	if (0 == ulLength)
	{
		return pstr;
	}
	ULONG ul = 0;
	for (; ul < ulLength-1; ul++)
	{
		CHAR tValue = *((*pdrgsz)[ul]);
		pstr->AppendFormat(GPOS_WSZ_LIT("%c%ls"), tValue, CDXLTokens::PstrToken(EdxltokenComma)->Wsz());
	}
	CHAR tValue = *((*pdrgsz)[ul]);
	pstr->AppendFormat(GPOS_WSZ_LIT("%c"), tValue);
```

It is as succinct as the original one. I will keep it as it is.

bhuvnesh2703 commented on libgpopt/src/translate/CTranslatorExprToDXL.cpp:
```diff
@@ -4702,17 +4710,42 @@ CTranslatorExprToDXL::ConstructLevelFiltersPartitionSelectorRange
-							NULL /*pmdidCastFunc*/,
-							ulLevel
-							);
+			if (fRangePart)
```
The name of the function is `ConstructLevelFiltersPartitionSelectorRange` which does not correlated with the new changes..

hsyuan replied:
hmm, how about `ConstructLevelFilters4PartitionSelector`?

bhuvnesh2703 commented:
LGTM (Bhuvnesh & Ekta)

hsyuan commented:
@d @bhuvnesh2703 @oarap @dhanashreek89, I think I addressed all the PR comments. Please let me know if you have any other comments / concerns. It passed all the test and hence I am thinking of pushing this tomorrow Morning. Please let me know if you want more time to review and I am more than happy to wait for it.

hsyuan commented:
Pushed to master, thanks all for reviewing!
